src/js/node/dgram.ts

```diff
@@ -1,23 +1,1082 @@
+      callback = port;
+      port = offset;
+      address = length;
+    }
+  } else {
+    if (typeof length === "number") {
+      buffer = sliceBuffer(buffer, offset, length);
+      if (typeof port === "function") {
+        callback = port;
+        port = null;
+      }
+    } else {
+      callback = offset;
+    }
+
+    if (port || address) throw new ERR_SOCKET_DGRAM_IS_CONNECTED();
+  }
+
+  if (!Array.isArray(buffer)) {
+    if (typeof buffer === "string") {
+      list = [Buffer.from(buffer)];
+    } else if (!ArrayBuffer.isView(buffer)) {
+      throw new ERR_INVALID_ARG_TYPE("buffer", ["Buffer", "TypedArray", "DataView", "string"], buffer);
+    } else {
+      list = [buffer];
+    }
+  } else if (!(list = fixBufferList(buffer))) {
+    throw new ERR_INVALID_ARG_TYPE("buffer list arguments", ["Buffer", "TypedArray", "DataView", "string"], buffer);
+  }
+
+  if (!connected) port = validatePort(port, "Port", false);
+
+  // Normalize callback so it's either a function or undefined but not anything
+  // else.
+  if (typeof callback !== "function") callback = undefined;
+
+  if (typeof address === "function") {
+    callback = address;
+    address = undefined;
+  } else if (address != null) {
+    validateString(address, "address");
+  }
+
+  if (state.bindState === BIND_STATE_UNBOUND) this.bind({ port: 0, exclusive: true }, null);
+
+  if (list.length === 0) list.push(Buffer.alloc(0));
+
+  // If the socket hasn't been bound yet, push the outbound packet onto the
+  // send queue and send after binding is complete.
+  if (state.bindState !== BIND_STATE_BOUND) {
+    enqueue(this, FunctionPrototypeBind(this.send, this, list, port, address, callback));
+    return;
+  }
+
+  const afterDns = (ex, ip) => {
+    defaultTriggerAsyncIdScope(this[async_id_symbol], doSend, ex, this, ip, list, address, port, callback);
+  };
+
+  if (!connected) {
+    state.handle.lookup(address, afterDns);
+  } else {
+    afterDns(null, null);
+  }
+};
+
+function doSend(ex, self, ip, list, address, port, callback) {
+  const state = self[kStateSymbol];
+
+  if (ex) {
+    if (typeof callback === "function") {
+      process.nextTick(callback, ex);
+      return;
+    }
+
+    process.nextTick(() => self.emit("error", ex));
+    return;
+  }
+  if (!state.handle) {
+    return;
+  }
+  const socket = state.handle.socket;
+  if (!socket) {
+    return;
+  }
+
+  let err = null;
+  let sent = 0;
+  try {
+    if (list.length === 1) {
+      if (port) {
+        socket.send(list[0], port, ip);
+      } else {
+        socket.send(list[0]);
+      }
+      sent = 1;
+    } else {
+      const packets = [];
+      if (port) {
+        for (const buf of list) {
+          packets.push(buf, port, address);
+        }
+      } else {
+        for (const buf of list) {
+          packets.push(buf);
+        }
+      }
+      sent += socket.sendMany(packets);
+    }
+  } catch (e) {
+    err = e;
+  }
+  // TODO check if this makes sense
+  if (callback) {
```

Comment:
the callback should only be called after drain has drained them all? or is that not how this works?